addons/binding/org.openhab.binding.robonect/src/main/java/org/openhab/binding/robonect/handler/RobonectHandler.java

```diff
@@ -0,0 +1,463 @@
+        }
+        if (error.getDate() != null) {
+            State dateTime = convertUnixToDateTimeType(error.getUnix());
+            updateState(CHANNEL_ERROR_DATE, dateTime);
+        }
+    }
+
+    private void updateNextTimer(MowerInfo info) {
+        State dateTime = convertUnixToDateTimeType(info.getTimer().getNext().getUnix());
+        updateState(CHANNEL_TIMER_NEXT_TIMER, dateTime);
+    }
+    
+
+    private State convertUnixToDateTimeType(String unixTimeSec) {
+        Instant ns = Instant.ofEpochMilli(Long.valueOf(unixTimeSec) * 1000);
+        ZonedDateTime zdt = ZonedDateTime.ofInstant(ns, ZoneId.of("UTC"));
+        return new DateTimeType(zdt);
+    }
+
+    private void refreshVersionInfo() {
+        VersionInfo info = robonectClient.getVersionInfo();
+        if (info.isSuccessful()) {
+            Map<String, String> properties = editProperties();
+            properties.put(Thing.PROPERTY_SERIAL_NUMBER, info.getRobonect().getSerial());
+            properties.put(Thing.PROPERTY_FIRMWARE_VERSION, info.getRobonect().getVersion());
+            properties.put(PROPERTY_COMPILED, info.getRobonect().getCompiled());
+            properties.put(PROPERTY_COMMENT, info.getRobonect().getComment());
+            updateProperties(properties);
+        } else {
+            logger.debug("Could not retrieve mower version info. Robonect error response message: {}",
+                         info.getErrorMessage());
+        }
+    }
+
+    private void refreshLastErrorInfo() {
+        ErrorList errorList = robonectClient.errorList();
+        if (errorList.isSuccessful()) {
+            if (errorList.getErrors() != null && errorList.getErrors().size() > 0) {
+                ErrorEntry lastErrorEntry = errorList.getErrors().get(0);
+                updateLastErrorChannels(lastErrorEntry);
+            }
+        } else {
+            logger.debug("Could not retrieve mower error list. Robonect error response message: {}",
+                         errorList.getErrorMessage());
+        }
+    }
+
+    private void updateLastErrorChannels(ErrorEntry error) {
+        if (error.getErrorMessage() != null) {
+            updateState(CHANNEL_LAST_ERROR_MESSAGE, new StringType(error.getErrorMessage()));
+        }
+        if (error.getErrorCode() != null) {
+            updateState(CHANNEL_LAST_ERROR_CODE, new DecimalType(error.getErrorCode().intValue()));
+        }
+        if (error.getDate() != null) {
+            State dateTime = convertUnixToDateTimeType(error.getUnix());
+            updateState(CHANNEL_LAST_ERROR_DATE, dateTime);
+        }
+    }
+
+    @Override
+    public void initialize() {
+        RobonectConfig robonectConfig = getConfigAs(RobonectConfig.class);
+        RobonectEndpoint endpoint = new RobonectEndpoint(robonectConfig.getHost(), robonectConfig.getUser(),
+                                                         robonectConfig.getPassword());
+        httpClient = new HttpClient();
```

Comment:
This is something @sjka already mentions. But after mentioning that I had to customize the authentication store of the HttpClient instance, he was also skeptical if this is gonna be a good idea.

Concrete I do the following in RobonectClient to get Preemptive Authentication working

```java
AuthenticationStore auth = httpClient.getAuthenticationStore();
        URI uri = URI.create(baseUrl);
        auth.addAuthenticationResult(new BasicResult(HttpHeader.AUTHORIZATION, uri, "Basic " + B64Code
                .encode(endpoint.getUser() + ":" + endpoint.getPassword(), StandardCharsets.ISO_8859_1)));
```

Do you think this will not have a side effect on others using this client? theoretically the authorizationResult is stored together with the URL, thus it could work..

During 2.2 development this was requrired, as the version of jetty client was quite old. Did jetty client version change since 2.2.0?

Reply:
No, Jetty is still 9.3.15(+) - so yes, I agree that it is probably safer to keep a separate client instance.
Nonetheless, you might want to retrieve it from `HttpClientFactory.createHttpClient()` as this will use a smaller thread pool (that can be configured for the system) and named threads, so that it is easier to find them in debugging sessions.

Reply:
ok, refactored to use factory